packages/dd-trace/test/plugins/agent.js

```diff
@@ -234,36 +234,37 @@ let availableEndpoints = DEFAULT_AVAILABLE_ENDPOINTS
  * @param {Set} [handlers] - Set of handlers to add the callback to.
  * @returns {Promise<void>} A promise resolving if expectations are met
  */
-function runCallback (callback, options, handlers) {
-  const deferred = {}
-  const promise = new Promise((resolve, reject) => {
-    deferred.resolve = resolve
-    deferred.reject = reject
+function runCallbackAgainstTraces (callback, options, handlers) {
+  let error
+
+  let resolve
+  let reject
+  const promise = new Promise((res, rej) => {
+    resolve = res
+    reject = rej
   })
 
-  const timeoutMs = options !== null && typeof options === 'object' && options.timeoutMs ? options.timeoutMs : 1000
+  const rejectionTimeout = setTimeout(() => {
+    if (error) reject(error)
+  }, options?.timeoutMs || 1000)
 
-  const timeout = setTimeout(() => {
-    if (error) {
-      deferred.reject(error)
-    }
-  }, timeoutMs)
-
-  let error
-  const handlerPayload = { handler, spanResourceMatch: options && options.spanResourceMatch }
+  const handlerPayload = {
+    handler,
+    spanResourceMatch: options?.spanResourceMatch
+  }
 
   function handler () {
     try {
       const result = callback.apply(null, arguments)
       handlers.delete(handlerPayload)
-      clearTimeout(timeout)
-      deferred.resolve(result)
+      clearTimeout(rejectionTimeout)
+      resolve(result)
     } catch (e) {
-      if (options && options.rejectFirst) {
-        clearTimeout(timeout)
-        deferred.reject(e)
+      if (options?.rejectFirst) {
+        clearTimeout(rejectionTimeout)
+        reject(e)
       } else {
-        error = error || e
+        error = error || e // if no spans match we report exactly the first mismatch error (which is unintuitive)
```

Comment:
We can use an AggregateError() instead

Reply:
I'm not even sure that's a good solution...

Basically the problem is that if there are three spans we're examining, A + B + C, and we're writing a test against span B, but we expect some field that isn't perfectly equal (so really we're looking for B' instead of B), the problem is that we say we don't match A.

Really what we want to know to help debug is information about the closest matching span, B in this case. An aggregate error shows us mismatched information about A and B and C which is a little better but still not perfect.

If we do switch to an approach like you're suggesting where we declaratively compare deep objects then we could calculate which is the "closest span" and display that specific span's callback error.